esr/esr_test.go

```diff
@@ -0,0 +1,614 @@
+				r = httptest.NewRequest(http.MethodPut, "http://localhost/reg", body)
+				r.Header = map[string][]string{"Content-Type": {"application/json"}}
+				return w, r
+			},
+			"Bad case, can't read body",
+		},
+		{
+			http.StatusBadRequest,
+			func(ua *UnitAsset) (w *httptest.ResponseRecorder, r *http.Request) {
+				ua.leading = true
+				w = httptest.NewRecorder()
+				body := io.NopCloser(strings.NewReader(""))
+				r = httptest.NewRequest(http.MethodPut, "http://localhost/reg", body)
+				r.Header = map[string][]string{"Content-Type": {"application/json"}}
+				return w, r
+			},
+			"Bad case, can't unpack body",
+		},
+		{
+			http.StatusInternalServerError,
+			func(ua *UnitAsset) (w *httptest.ResponseRecorder, r *http.Request) {
+				ua.leading = true
+
+				w = httptest.NewRecorder()
+
+				// Record has to match the one sent by sendAddRequest(..)
+				rec := &forms.ServiceRecord_v1{
+					Id:                int(0),
+					ServiceDefinition: "test",
+					SystemName:        "System",
+					ServiceNode:       "node",
+					IPAddresses:       []string{"123.456.789.012"},
+					ProtoPort:         map[string]int{"http": 1234},
+					Details:           map[string][]string{"details": {}},
+					Certificate:       "ABCD",
+					SubPath:           "testPath",
+					RegLife:           25,
+					Version:           "SignalA_v1.0",
+					Created:           "",
+					Updated:           time.Now().String(),
+					EndOfValidity:     time.Now().Add(25 * time.Second).String(),
+					SubscribeAble:     false,
+					ACost:             float64(0),
+					CUnit:             "",
+				}
+				data, err := json.Marshal(rec)
+				if err != nil {
+					log.Printf("Error: %v", err)
+				}
+				body := io.NopCloser(bytes.NewReader(data))
+				r = httptest.NewRequest(http.MethodPut, "http://localhost/reg", body)
+				r.Header = map[string][]string{"Content-Type": {"application/json"}}
+				sendAddRequest(0, "test", "testPath", "", ua.requests)
+
+				return w, r
+			},
+			"Bad case, request returns error",
+		},
+		{
+			200,
+			func(ua *UnitAsset) (w *httptest.ResponseRecorder, r *http.Request) {
+				ua.leading = true
+
+				rec := &forms.ServiceRecord_v1{
+					Id:      0,
+					Version: "ServiceRecord_v1",
+				}
+
+				data, _ := json.Marshal(rec)
+				w = httptest.NewRecorder()
+				body := io.NopCloser(bytes.NewReader(data))
+				r = httptest.NewRequest(http.MethodPost, "http://localhost/reg", body)
+				r.Header = map[string][]string{"Content-Type": {"application/json"}}
+
+				return w, r
+			},
+			"Good case, everything passes",
+		},
+		{
+			200,
+			func(ua *UnitAsset) (w *httptest.ResponseRecorder, r *http.Request) {
+				ua.leading = true
+				w = httptest.NewRecorder()
+				body := io.NopCloser(strings.NewReader(""))
+				r = httptest.NewRequest(http.MethodGet, "http://localhost/reg", body)
+				r.Header = map[string][]string{"Content-Type": {"application/json"}}
+
+				return w, r
+			},
+			"Good case, default case",
+		},
```

Comment:
These setup functions all looks very similar. Would be better to break out `ua.leading` and the body and use them in the table instead of the setup func. The two cases where you're doing something different in the setup could probably be refactored or handled as separate special cases.